Introducing the Pipeline class

mlange05 opened the pull request:
_Note: For regression testing purposes this is stacked on top of PR #259._

This PR introduces the first basic implementation of a `Pipeline` abstraction that allows multiple `Transformation` classes to be combined in a "pipeline", so that they're effect may be applied to single items or entire scheduler graphs in a single invocation. As a result, we now have a mechanism to recover and provide the high-level SCC-style family of transformations as a collection of `Pipeline` objects that behave like a single `Transformation` class. 

Each of these pilelines ("SCCvector", "SCCHoist", "SCCStack") is now composed of the shared and distinct sub-transformations, and captures and manages the respective constructor arguments for each transformation. This step required moving some of the sub-transformations into separate sub-packages to break circular import dependencies.

For this PR, I have only focused on the basics of providing the necessary mechanics and updating the use of the SCC transformations in `loki-transform.py`. More additions, like `pipelin.append()/.prepend()` or configuration of pipelines from the config files are left for future developments.

In some more detail:
* New `Pipeline` class that manages kw-arguments in it's constructor and dispatches to the respective `Transformation` classes upon instantiation
* Added `process_transformation` and `process_pipeline` to `Scheduler` and dispatch to the respective variant in `process()`
* Re-organise SCC components to break cyclic import dependencies
* Drop explicit managing of `key` as a constructor argument from the hoisting classes
* Provide `SCCVectorPipeline`, `SCCHoistPipeline` and `SCCStackPipeline` as partial pipeline constructors and use in `loki-transform.py`, as well as respective tests.

mlange05 commented:
For the record, EC-physics regression seems to work with this.

mlange05 commented on loki/transform/pipeline.py:
```diff
@@ -0,0 +1,74 @@
+        for cls in classes:
+
+            # Get all relevant constructor parameters from teh MRO,
+            # but exclude catch-all kwyward args, like ``**kwargs``
```
typo

mlange05 commented on loki/batch/scheduler.py:
```diff
@@ -377,6 +377,46 @@ def rekey_item_cache(self):
+            The transformation pipeline to apply
+        """
+        for transformation in pipeline.transformations:
+            self.process_transformation(transformation)
```
No, in the follow-up PR #271 we introduce native addition of pipelines and transformations, and they always result in a new pipeline, eg. `pipeline = short_pipeline + other_pipeline + some_trafo` will result in a single `Pipeline` object that encapsulates a single sequence of `Transformation` objects by appending them as the notation suggests. 

A "pipeline of pipelines" could potentially become problematic if we extend the abstraction with further mechanics, so I'd like to avoid that. We could maybe make "list of Pipeline" a thing, but I feel that's beyond the intro PR. :wink:

mlange05 commented on scripts/loki_transform.py:
```diff
@@ -225,66 +223,73 @@ def convert(
+
+        scheduler.process( NormalizeRangeIndexingTransformation() )
+
+    if global_var_offload:
+        scheduler.process(transformation=GlobalVariableAnalysis())
+        scheduler.process(transformation=GlobalVarOffloadTransformation())
```
I wondered the same thing, but CI seemed happy. Paging @awnawab for confirmation though?

awnawab replied:
I can't think of an example where the SCC family of trafos would change the offload requirements. Plus if the CI is happy, then even with the changes I have in the pipeline we should be in good shape 👍

mlange05 commented:
@reuterbal I Fixed the typos and removed the key as suggested. Please feel free to rebase if you want to see the missing coverage report. 

And yes, the re-jigging of the SCC components was mechanic. Any further clean-up there will come via follow-on PRs.
